text/0225-configuration.md

```diff
@@ -0,0 +1,329 @@
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method then applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
+
+### ParseAndValidateConfigurationFromFile(filepath, format) -> config
+
+An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. YAML *SHOULD* be supported, but JSON, which is a strict subset of YAML, *MAY* be supported if a language ecosystem has good reason. For example a language that has built-in support for JSON running in an environment that is sensitive to additional dependencies *MAY* only be able to support JSON configuration.
+
+The method returns a `Configuration` model that has been validated. This API *MAY* return an error or raise an exception, whichever is idiomatic to the implementation for the following reasons:
+
+* file doesn't exist or is invalid
+* configuration parsed is invalid
+
+#### Python ParseAndValidateConfigurationFromFile example
+
+```python
+
+filepath = "./config.yaml"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filename)
+except Exception as e:
+  print(e)
+
+filename = "./config.json"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filename, format="json")
+except Exception as e:
+  raise e
+
+```
+
+#### Go ParseAndValidateConfigurationFromFile example
+
+```go
+
+filename := "./config.yaml"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename)
+if err != nil {
+  return err
+}
+
+filename := "./config.json"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename, otelconfig.WithFormat("json"))
+if err != nil {
+  return err
+}
+
+```
+
+Implementations *MUST* allow users to specify an environment variable to set the configuration file. This gives flexibility to end users of implementations that do not support command line arguments. The proposed name for this variable:
+
+* `OTEL_CONFIG_FILE`
```

Comment:
Should the implementation choose YAML or JSON using file extension or by inspecting the contents of the file? Or is this only meant to support whatever the default file format is?

Reply:
I think if an implementation supports both they should choose the correct one based on extension. If an implementation only supports one format (i.e. JSON) and `config.yaml` is specified, that implementation should throw an error. It can produce the error either based on detecting the incorrect extension, or by trying to parse the file and waiting for things to blow up 😂.

Reply:
Added a note to call out that the file extension will be used to detect the format.